### PR TITLE
add group anagram algorithm

### PR DESCRIPTION
**Pull request by Kishore-abhimanyu**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: strings/group_anagram.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import doctest
+
+
+def find_anagrams(_list: list) -> list[list[str]]:
```
[the remark this answers is not in the document]
Comment (Kishore-abhimanyu) [reply strings/group_anagram.py:4]: I have added the doctest

**Comment by Kishore-abhimanyu** [comment]
I have resolved this issue

**Review by CaedenPH — CHANGES_REQUESTED** [review]
Please can you add more comments and docstrings so the reader understands what is happening at all times

**Comment by Kishore-abhimanyu** [comment]
> Please can you add more comments and docstrings so the reader understands what is happening at all times

Comment are added

File: strings/group_anagram.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
```
[the remark this answers is not in the document]
Comment (Kishore-abhimanyu) [reply strings/group_anagram.py:4]: doctest is also added with the name group_anagram.txt

**Comment by Kishore-abhimanyu** [comment]
Can you check now

File: strings/group_anagram.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            if (
+                char_map[_keys[itr1]] == char_map[_keys[itr2]]
+                and char_map[_keys[itr2]]["is_visited"] == 0
+            ):
+                res.append(_keys[itr1])
+                res.append(_keys[itr2])
+                # once the word is compared and match is found it is marked as visited
+                char_map[_keys[itr2]]["is_visited"] = 1
+        # now sort the words and remove the duplicates if present
+        _set = sorted(list(set(res)))
```
Comment (cclauss) [inline strings/group_anagram.py:39]:
```suggestion
        _set = sorted(set(res))
```

File: strings/group_anagram.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            # if it is already present add the count of each letter
+            if char not in char_map[each]:
+                char_map[each][char] = 1
+            else:
+                char_map[each][char] += 1
+
+    """
+    Match each words with the occurrence count
+    """
+    _keys = list(char_map.keys())
```
Comment (cclauss) [inline strings/group_anagram.py:23]:
```suggestion
    _keys = list(char_map)
```

File: strings/group_anagram.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cclauss) [reply strings/group_anagram.py:4]: This function does NOT have a doctes.

File: strings/group_anagram.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def find_anagrams(_list: list) -> list[list[str]]:
+    """
+    Find the characters and number of occurrence in each words
+    """
+    char_map = {}
+    anagrams = []
+    for each in _list:
+        char_map[each] = {"is_visited": 0}
```
Comment (cclauss) [inline strings/group_anagram.py:11]:
```suggestion
        char_map[each] = {"is_visited": False}
```

File: strings/group_anagram.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    _keys = list(char_map.keys())
+    for itr1 in range(0, len(_keys)):
+        # One word is taken at a time
+        res = []
+        for itr2 in range(itr1 + 1, len(_keys)):
+            # now the character count of each work is compared with the main word
+            # if the character count is matched it is considered as an anagram
+            if (
+                char_map[_keys[itr1]] == char_map[_keys[itr2]]
+                and char_map[_keys[itr2]]["is_visited"] == 0
```
Comment (cclauss) [inline strings/group_anagram.py:32]:
```suggestion
                and not char_map[_keys[itr2]]["is_visited"]
```

File: strings/group_anagram.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            ):
+                res.append(_keys[itr1])
+                res.append(_keys[itr2])
+                # once the word is compared and match is found it is marked as visited
+                char_map[_keys[itr2]]["is_visited"] = 1
```
Comment (cclauss) [inline strings/group_anagram.py:37]:
```suggestion
                char_map[_keys[itr2]]["is_visited"] = True
```

File: strings/group_anagram.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        _set = sorted(list(set(res)))
+
+        # if there are anagrams, push it to the res array
+        if len(_set) != 0:
+            anagrams.append(_set)
+        # if there are no anagrams found return []
+    return anagrams
+
+
+def test_find_anagrams() -> None:
```
Comment (cclauss) [inline strings/group_anagram.py:48]: Put the test in the function so that if someone copies the function into their code, the test comes with it.

**Comment by Kishore-abhimanyu** [comment]
The changes has been updated. Could you check the changes

File: strings/group_anagram.py [hunk 9]
```diff
@@ -9,7 +9,8 @@ def find_anagrams(_list: list) -> list[list[str]]:
     char_map = {}
     anagrams = []
     for each in _list:
-        char_map[each] = {"is_visited": False}
+        char_map[each] = {}
+        char_map[each]["is_visited"] = False
```
Comment (cclauss) [inline strings/group_anagram.py:13]: Why is this an improvement?

Comment (Kishore-abhimanyu) [reply strings/group_anagram.py:13]: If char_map[each] is responsible for both tacking the visitors and the count, if it is referenced before assignment, there may be a cause of runtime error

**Comment by cclauss** [comment]
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

strings/group_anagram.py:9: error: Need type annotation for "char_map" (hint: "char_map: Dict[<type>, <type>] = ...")

**Comment by cclauss** [comment]
Please read CONTRIBUTING.md.

Hacktoberfest is strenuous for our team of active volunteer maintainers. That small team now has ___200+ open pull requests___ to review.  We have __high standards for our submissions__ which are described in [`CONTRIBUTING.md`](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#contributing-guidelines).  Our continuous integration tests help to ensure that our standards are met before we review and merge deserving pull requests.  This is further complicated by the fact that we already have [lots of algorithms](https://github.com/TheAlgorithms/Python/blob/master/DIRECTORY.md) and would like to avoid duplicates.

For these reasons, I will **close all open pull requests failing tests** so that our unpaid team has a smaller mountain to climb.  If you want us to land your pull request, please read [`CONTRIBUTING.md`](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#contributing-guidelines) _carefully_ and try again.

**Pull request closed without merging**